new JLogger type: JLoggerCallback (issue #866)

okonomiyaki3000 opened the pull request:
This adds a new JLogger type that simply passes a log entry to a user defined callback function. Callback can be in any of the callback formats normally used by PHP, including lambda (available since PHP 5.3).

In order to properly support lambda functions (Closure objects), a minor modification to JLog was required.

okonomiyaki3000 committed:
added support for new JLogger type: JLoggerCallback

okonomiyaki3000 committed:
added support for new JLogger type: JLoggerCallback

mbabker commented on libraries/joomla/log/log.php:
```diff
@@ -13,7 +13,7 @@
 JLoader::register('LogException', JPATH_PLATFORM . '/joomla/log/logexception.php');
 
-JLoader::discover('JLogger', __DIR__ . '/loggers');
+JLoader::discover('JLogger', dirname(__FILE__) . '/loggers');
```
Platform 12.1 requires PHP 5.3 minimum, and `__DIR__` is valid in 5.3.  Can you change this back?

mbabker commented on libraries/joomla/log/log.php:
```diff
@@ -419,4 +433,28 @@ protected function findLoggers($priority, $category)
+	 *
+	 * @return  string  A unique identifier for the passed object
+	 *
+	 * @since   11.5
```
`@since` tags should be for 12.1 for new development.  There won't be a 11.5 release.

mbabker commented on libraries/joomla/log/log.php:
```diff
@@ -419,4 +433,28 @@ protected function findLoggers($priority, $category)
+			$id = count($closures) - 1;
+		}
+		
+		return '_CLOSURE_ID_'.$id;
```
To match the code style rules, this line should be `return '_CLOSURE_ID_' . $id;`

okonomiyaki3000 committed:
changes for style, versioning, etc.

okonomiyaki3000 commented:
OK, that should do it.

joomla-jenkins commented:
Unit testing complete.  There were 0 failures and 0 errors from 2012 tests and 11224 assertions.
Checkstyle analysis reported 164 warnings and 9 errors.

okonomiyaki3000 commented:
Where can I see details of this checkstyle analysis?

realityking commented:
You have to run phpcs on your own computer. The code style definition is in the build/phpcs folder.

okonomiyaki3000 committed:
fixes to coding style. white-space, etc.

okonomiyaki3000 commented:
Fixed the 9 errors but phpcs didn't turn up any warnings for me.

joomla-jenkins commented:
Build triggered by changes to the head.

Unit testing complete.  There were 0 failures and 10 errors from 2023 tests and 11257 assertions.
Checkstyle analysis reported 164 warnings and 0 errors.

eddieajau commented:
Just trying to understand what's being proposed.  Why would you use a callback function when you can just as easily make your own logger class?

okonomiyaki3000 commented:
It is surely simpler to write a function than a class. Especially with the lambda support in PHP 5.3, you can get a lot of versatility with very little effort. Maybe I've written too much Javascript but I really like the idea of extending functionality by just passing a function. 

I imagine that someone might use this in case their Joomla! installation is part of a larger ecosystem of apps and they'd like to hand off logging to some other part of that system. Or perhaps during some testing phase you find that you need some specific logging functionality on a temporary basis. Of course, you can write a class for these cases but just passing a function may be faster/easier/more fun?

eddieajau commented:
Ok, fair enough.  I'm not sure I'd use it, but I can see where you are coming from.  You have some testing errors to fix and we'd want some additional unit testing to verify it all works as advertised :)

eddieajau commented:
Oh, and if you can give me some explanatory text, I'll put it in the dev manual (http://developer.joomla.org/manual/chap-Joomla_Platform_Manual-Log.html) unless you want to have a crack at that as well.

Thanks in advance.

okonomiyaki3000 committed:
Merge remote-tracking branch 'upstream/staging' into staging

okonomiyaki3000 committed:
Merge branch 'master' of https://github.com/joomla/joomla-platform into JLoggerCallback

okonomiyaki3000 commented:
I don't think these testing errors have anything to do with me. I get the same errors whether I run the tests on master or staging. Shouldn't master run without errors if I haven't touched it?

joomla-jenkins commented:
Build triggered by changes to the head.

Unit testing complete.  There were 0 failures and 0 errors from 2193 tests and 11591 assertions.
Checkstyle analysis reported 163 warnings and 0 errors.

robschley commented on libraries/joomla/log/loggers/callback.php:
```diff
@@ -0,0 +1,69 @@
+		// Pass the log entry to the callback function
+		if (!call_user_func($this->callback, $entry))
+		{
+			throw new LogException;
```
There should be a message for every exception thrown.

okonomiyaki3000 replied:
In fact, I think we don't need an exception here at all. I put it in because the FormattedText logger throws one (also without a message) but, on second thought, it doesn't make much sense in this case. Any problem to just remove it?

robschley commented:
I think that if we're going to add this we really need some tests for it. I suspect this is the type of feature that not many people will use so if there are no tests it is likely to get neglected.

okonomiyaki3000 committed:
Merge branch 'staging' of https://github.com/joomla/joomla-platform into staging

okonomiyaki3000 committed:
Merge branch 'staging' into JLoggerCallback

okonomiyaki3000 committed:
Just execute the callback. Ignore return values. Callback is not required to return 'true'

okonomiyaki3000 committed:
Throw an exception if there is not a valid callback.

okonomiyaki3000 committed:
Tests for JLoggerCallback

okonomiyaki3000 commented:
I've also written some documentation for this that should be suitable for the manual. What should I do with that?

joomla-jenkins commented:
Build triggered by changes to the head.

Unit testing complete.  There were 0 failures and 0 errors from 2204 tests and 11608 assertions.
Checkstyle analysis reported 163 warnings and 0 errors.

okonomiyaki3000 commented:
Hey guys, anything happening here?

chdemko commented:
The pull request cannot be merged (libraries/joomla/log/loggers has been renamed to libraries/joomla/log/logger)

okonomiyaki3000 committed:
Merge branch 'staging' of https://github.com/joomla/joomla-platform into staging

okonomiyaki3000 committed:
moved callback.php from /libraries/joomla/log/loggers/ to /libraries/joomla/log/logger/
Also moved related tests to the new location

okonomiyaki3000 committed:
Changes to JLoggerCallbackTest
- Look in the correct new directory for JLoggerCallback
- extend PHPUnit_Framework_TestCase instead of JoomlaTestCase

okonomiyaki3000 commented:
OK, all updated to the new locations

chdemko commented on libraries/joomla/log/log.php:
```diff
@@ -188,9 +188,23 @@ public static function addLogger(array $options, $priorities = self::ALL, $categ
+		if (isset($options['callback']) && is_a($options['callback'], 'closure'))
+		{
+			$callback = $options['callback'];
+			$options['callback'] = self::getClosureUniqueId($options['callback']);
```
You can use spl_hash_object($options['callback']); and drop the getClosureUniqueId function

chdemko commented on libraries/joomla/log/logger/callback.php:
```diff
@@ -0,0 +1,70 @@
+
+defined('JPATH_PLATFORM') or die;
+
+jimport('joomla.log.logger');
```
This is useless

chdemko replied:
Could you please remove this line?

chdemko commented on libraries/joomla/log/logger/callback.php:
```diff
@@ -0,0 +1,70 @@
+ * @subpackage  Log
+ * @since       12.1
+ */
+class JLoggerCallback extends JLogger
```
Use JLogLogger (JLogger is deprecated)

chdemko commented on libraries/joomla/log/logger/callback.php:
```diff
@@ -0,0 +1,70 @@
+ *
+ * @package     Joomla.Platform
+ * @subpackage  Log
+ * @since       12.1
```
12.2

chdemko commented on libraries/joomla/log/logger/callback.php:
```diff
@@ -0,0 +1,70 @@
+	 *
+	 * @param   array  &$options  Log object options.
+	 *
+	 * @since   12.1
```
12.2

chdemko commented on libraries/joomla/log/logger/callback.php:
```diff
@@ -0,0 +1,70 @@
+	 *
+	 * @return  boolean  True on success.
+	 *
+	 * @since   12.1
```
12.2

chdemko commented on tests/suites/unit/joomla/log/loggers/JLoggerCallbackTest.php:
```diff
@@ -0,0 +1,184 @@
+ *
+ * @package     Joomla.Platform
+ * @subpackage  Log
+ * @since       12.1
```
12.2

chdemko commented on tests/suites/unit/joomla/log/loggers/stubs/callback/helper.php:
```diff
@@ -0,0 +1,60 @@
+ *
+ * @package     Joomla.UnitTest
+ * @subpackage  Log
+ * @since       12.1
```
12.2

chdemko commented:
Please add @since 12.2 to all your test functions

okonomiyaki3000 committed:
Merge branch 'staging' of https://github.com/joomla/joomla-platform into staging

okonomiyaki3000 committed:
Rename JLoggerCallbackTest.php to JLogLoggerCallbackTest.php
Add @since 12.2 in all relevant cases
use spl_object_hash to get unique id for closures

okonomiyaki3000 committed:
extend JLogLogger, not JLogger which is now deprecated

okonomiyaki3000 committed:
Remove unnecessary jimport

chdemko commented:
Can you please make all references to 12.1 to be 12.2 and add <code>@since</code> to your test methods

okonomiyaki3000 committed:
Added @since 12.2 to a few test functions that were missing it
